### PR TITLE
Allow custom association between file names and grammars

### LINKED ISSUE #1718: Make file type -> language mapping locally configurable
Several halp issues about people wanting to add extensions to language plugins.

Currently they have to fork the package to do this, might be nice to support this through the config.

### PR DESCRIPTION
**Pull request by maxbrunsfeld**
Closes #1718

### CODE REVIEW

**Comment by benogle** [comment]
Cool! People will be really happy with this.

**Comment by maxbrunsfeld** [comment]
Right now, the config key is `core.fileTypesByScope`, and works like so:

```coffee
'*':
  core:
    fileTypesByScope:
      'source.ruby': ['Cheffile', '.also-ruby']
      'source.cpp': ['h']
```

What do you all think @atom/feedback?

**Comment by thomasjo** [comment]
> Right now, the config key is `core.fileTypesByScope` [..]

Not falling in love with the name, but other than that this is awesome. The name might grow on me.

File: src/config.coffee [hunk 1]
```diff
@@ -665,12 +665,23 @@ class Config
...
-      schema = schema.properties?[key]
+      if schema.type is 'object'
+        childSchema = schema.properties?[key]
+        unless childSchema?
+          if isPlainObject(schema.additionalProperties)
+            childSchema = schema.additionalProperties
+          else if schema.additionalProperties is false
+            return false
+          else
+            return null
```
Comment (as-cii) [inline src/config.coffee]: I am not 100% familiar with this code but I am wondering if we could avoid `null` and/or `false` as this function's return values. Without looking at the documentation, the calling code seems a bit obscure to me.

I am probably ok with keeping `null`, as it sounds reasonable that a schema cannot be found. But can we encode `false` differently? Can we tell what that value represents without indicating it in the documentation?

Maybe we could make the return value consistent across the various `return` statements, and return an object that always responds to the following messages:

* `hasSchemaSpecified`
* `isAccessibleFromRootSchema`

@maxbrunsfeld: what do you think?

**Comment by as-cii** [comment]
One comment, otherwise it looks good to me! Great job, @maxbrunsfeld! :sparkles:

**Comment by maxbrunsfeld** [comment]
> Not falling in love with the name

Yeah, I feel that. `fileTypesByScope` communicates the structure of the object pretty exactly, but maybe it would be better to give it a simpler name like `customFiletypes` and leave it up to the description to describe the name.

File: src/config.coffee (hunk 1 above)
Comment (maxbrunsfeld) [inline src/config.coffee]: I agree that the `null/false` thing is weird. But I feel like there's something intuitive about  `::getSchema` returning the same thing that was passed to `::setSchema`, as opposed to an object with additional structure. Maybe we could think of another way to encode the difference between "not accessible" and "whatever value you want" though.

How about we change `null` to represent "not accessible", and in the case where any value is allowed (e.g. `atom.config.getSchema("editor.unknownKey")`), we return a new schema like `{type: "any"}` that corresponds to no validation whatsoever.

**Comment by benogle** [comment]
> Right now, the config key is core.fileTypesByScope, and works like so:

Any reason these couldnt be a minimatch pattern. Maybe overkill, but could be a little more flexible:

```coffee
'*':
  core:
    fileTypesByScope:
      'source.ruby': ['Cheffile', '*.this-is-also-ruby']
      'source.cpp': ['*.h']
      'source.js': ['i-am-crazy*.*']
```

**Comment by benogle** [comment]
For the record, I dig the name for its clarity.

**Comment by maxbrunsfeld** [comment]
> Any reason these couldn't be a minimatch pattern.

I wanted to make them use the same system as the `grammars` use for matching files ([exact suffix matches](https://github.com/atom/language-coffee-script/blob/master/grammars/coffeescript.cson#L3)), so it would be easiest to understand which matches take precedence. Right now, longer matches win, and ties are broken by choosing the users' custom filetypes.

**Comment by benogle** [comment]
> I wanted to make them use the same system as packages' grammars use

:+1:

File: src/config.coffee (hunk 1 above)
Comment (as-cii) [inline src/config.coffee]: > Another thing we could do is change null to represent "not accessible", and in the case where any value is allowed (e.g. atom.config.getSchema("editor.unknownKey")), we return a new schema like {type: "any"} that corresponds to no validation whatsoever.

:+1:

**Comment by maxbrunsfeld** [comment]
Ok, I'm leaning toward renaming it from `fileTypesByScope` to `customFiletypes`. Anybody care?

**Comment by nathansobo** [comment]
I like `customFileTypes`... less precise but more descriptive somehow.

**Comment by maxbrunsfeld** [comment]
Ok, I think this is ready. Anywhere else I should document this new setting?

**Comment by kevinsawicki** [comment]
nice ship :ship: :sunglasses:

**Comment by rubidot** [comment]
For anyone looking for how to use this featured: [Customizing Language Recognition](http://flight-manual.atom.io/using-atom/sections/basic-customization/#customizing-language-recognition)

**Pull request merged**